core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/BavetConstraintSessionFactory.java

```diff
@@ -140,6 +141,13 @@ private long determineLayerIndex(AbstractNode node, NodeBuildHelper<Score_> buil
             var leftParentNode = buildHelper.findParentNode(leftParent);
             var rightParentNode = buildHelper.findParentNode(rightParent);
             return Math.max(leftParentNode.getLayerIndex(), rightParentNode.getLayerIndex()) + 1;
+        } else if (node instanceof AbstractConcatNode<?> concatNode) {
+            var nodeCreator = (BavetJoinConstraintStream<?>) buildHelper.getNodeCreatingStream(concatNode);
```

Comment:
Why would concat implement join? The two operations have barely anything in common.

Reply:
Largely for a fact that like `join`, `concat` has two parents, and the only two methods on `BavetJoinConstraintStream` are getters for the two parents. IMO, it would be pointless for two classes to have exactly the same methods, although those two methods could be extracted into a super interface `BavetBinaryOperatorResultStream` and have `BavetJoinConstraintStream` extend that (essentially making `BavetJoinConstraintStream` a marker interface).